### ISSUE TITLE
A better way to store extension data on the environment

### ISSUE THREAD

**Issue by timhoffm** [sphinx-doc/sphinx#9003]
### Problem

The approach to storing extension data is explained only vaguely.

From the [env-merge-info description](https://www.sphinx-doc.org/en/master/extdev/appapi.html#event-env-merge-info) and 
the [todo extension tutorial](https://www.sphinx-doc.org/en/master/development/tutorials/todo.html) I infer that one has to add arbitrary attributes to the `env` object.

This doesn't look very clean. It risks breaking the object and one has to be careful that the data is already added (lots of `hasattr()` checks).

### Proposal

- Add a dict `env.extension_data` and recommend extensions writing there.
- Can one initialize the extension data instead of checking in various places if it exists? Would `env-before-read-docs` the right place? I.e.
~~~
def setup(app):
    ...
    app.connect("env-before-read-docs", setup_data)

def setup_data(app, env, docnames):
    env.extension_data['todo_all_todos'] = []
~~~

**Comment by tk0miya** [comment]
Absolutely. Adding arbitrary attributes to the env object is ugly. Actually, the real "todo" extension uses a Domain component as a datastore. It can define methods for initialization, gathering contents, merging, and purging. To support parallel build, extension should support both merging and purging. So the Domain component is better to implement it, I think
